### PR TITLE
fix: Normalize endpoint and web _router_completed calls and tests

### LINKED ISSUE #549: Ensure router_type is present in _router_completed
A previous PR fixed the return handing in _router_completed, this fix was not applied to the _router_completed in web/webpush.py however.

### PR DESCRIPTION
**Pull request by jrconlin**
closes #549 

@bbangert r?

### CODE REVIEW

**Comment by bbangert** [comment]
Which way is it supposed to be? The endpoint.py change to remove the del statement was done first, so I had thought the endpoint.py change was going to propagate to webpush.py, not vice versa. Is it supposed to delete the router_data if its not set, or not?

**Comment by bbangert** [comment]
r+

File: autopush/router/fcm.py [hunk 1]
```diff
@@ -45,18 +45,19 @@ def amend_msg(self, msg, data=None):
...
-        # ensure that the client's senderid value matches what we need
-        # it to be. (If the client has an unexpected or invalid SenderID,
+        # senderid is the remote client's senderID value. This value is
+        # very difficult for the client to change, and there was a problem
+        # where some clients had an older, invalid senderID. We need to
+        # be able to match senderID to it's corresponding auth key.
+        # If the client has an unexpected or invalid SenderID,
         # it is impossible for us to reach them.
-        if not (router_token == router_data['token'] == self.senderID):
+        if not (senderid == self.senderID):
```
Comment (jrconlin) [inline autopush/router/fcm.py]: router['token'] is the registrationID and should NOT match senderid. I'm an idiot.

File: autopush/web/webpush.py [hunk 2]
```diff
@@ -56,13 +56,21 @@ def post(self, api_ver="v1", token=None):
...
+                # be deleted. There is no longer valid route information for
+                # this record.
+                self.ap_settings.router.drop_user(self.uaid)
+                return self._router_response(response)
+            # The router data needs to be updated to include any changes
+            # requested by the bridge system
+            uaid_data["router_data"] = response.router_data
+            # set the AWS mandatory data
             uaid_data["connected_at"] = ms_time()
+            rt = self.valid_input.get(
```
Comment (jrconlin) [inline autopush/web/webpush.py]: Aye! Dig! Dig ya miserable bastard! There be a router_type down there!

File: autopush/endpoint.py [hunk 3]
```diff
@@ -584,11 +584,21 @@ def _route_notification(self, version, result, data, ttl=None):
...
         # TODO: Add some custom wake logic here
-
         # Were we told to update the router data?
-        if response.router_data:
+        if response.router_data is not None:
+            if not response.router_data:
+                # An empty router_data object indicates that the record should
+                # be deleted. There is no longer valid route information for
+                # this record.
+                self.ap_settings.router.drop_user(self.uaid)
```
Comment (pjenvey) [inline autopush/endpoint.py]: I think we need to deferToThread drop_user

Comment (jrconlin) [inline autopush/endpoint.py]: I was wondering about that. In theory, we're already in a deferred thread from _route_notification.
I'll add a defer anyway.

File: autopush/router/gcm.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 class GCMRouter(object):
     """GCM Router Implementation"""
     log = Logger()
-    gcm = None
+    gcm = {}
```
Comment (pjenvey) [inline autopush/router/gcm.py]: We only instantiate one GCMRouter per app it looks like, but let's avoid this evil mutable class variable regardless

**Comment by jrconlin** [comment]
ready for re-review

@bbangert @pjenvey r?

File: autopush/router/gcm.py [hunk 5]
```diff
@@ -22,39 +21,45 @@ def __init__(self, ap_settings, router_conf):
         self.min_ttl = router_conf.get("ttl", 60)
         self.dryRun = router_conf.get("dryrun", False)
         self.collapseKey = router_conf.get("collapseKey", "simplepush")
-        self.senderIDs = router_conf.get("senderIDs")
+        self.gcm = {}
+        self.senderIDs = {}
```
Comment (pjenvey) [inline autopush/router/gcm.py]: You could actually combine the gcm and senderIDs dicts into one at this point, {sid: (auth, GCM()),} .. but wait a second..

**Comment by pjenvey** [comment]
r+

File: autopush/router/gcm.py (hunk 5 above)
Comment (jrconlin) [inline autopush/router/gcm.py]: well, really, I could create a map of {sid: GCM()} Holding onto the auth key is probably not needed. I divided things up for now mostly to keep it clear (at least for me). We can collapse things down if we need to.

File: autopush/web/webpush.py (hunk 2 above)
Comment (bbangert) [inline autopush/web/webpush.py]: Why not just record the router_type on self from the post method above that already dug it out?

Comment (bbangert) [inline autopush/web/webpush.py]: Also, the valid_input was guaranteed to have a router_type and subscription attribute already, the benefit of validation up front means we aren't guessing if it exists and using get calls like this.

File: autopush/web/webpush.py [hunk 6]
```diff
@@ -56,13 +56,23 @@ def post(self, api_ver="v1", token=None):
...
+                d.addCallback(lambda x: self._router_response(response))
+                return d
+            # The router data needs to be updated to include any changes
+            # requested by the bridge system
+            uaid_data["router_data"] = response.router_data
+            # set the AWS mandatory data
             uaid_data["connected_at"] = ms_time()
+            rt = self.valid_input.get(
+                'subscription', {}).get('user_data', {}).get('router_type')
+            uaid_data["router_type"] = uaid_data.get("router_type", rt)
```
Comment (bbangert) [inline autopush/web/webpush.py]: What's the rationale for a router being able to change the router type?

**Comment by bbangert** [comment]
r-, need some clarification on what's occuring with router type resets, and get calls that aren't needed.

File: autopush/web/webpush.py (hunk 6 above)
Comment (jrconlin) [inline autopush/web/webpush.py]: We're setting it if it's not already set.

Comment (bbangert) [inline autopush/web/webpush.py]: How would we get to this point if it wasn't set? The validation logic will toss out the request if there's no router_type before starting this process.

**Comment by bbangert** [comment]
r+ fe0d19c

**Pull request merged**